File: packages/zynth-async-storage/android/ZynthAsyncStorage/src/main/java/dev/zynth/asyncstorage/ZynthAsyncStorageModule.kt
package dev.zynth.asyncstorage

import android.content.Context
import android.content.SharedPreferences
import com.zynth.kit.runtime.ZynthModule
import com.zynth.kit.runtime.ZynthSyncModule
import com.zynth.kit.runtime.ZynthArgs
import org.json.JSONArray
import org.json.JSONObject

class ZynthAsyncStorageModule(
    context: Context
) : ZynthModule, ZynthSyncModule {
    override val name: String = "ZynthAsyncStorage"
    private val storage = ZynthAsyncStorageStore(context)

    override val exportedMethods: List<String> = listOf(
        "getItem", "setItem", "removeItem", "mergeItem", "clear",
        "getAllKeys", "multiGet", "multiSet", "multiRemove", "multiMerge"
    )

    override val protectedMethods: List<String> = listOf(
        "setItem", "removeItem", "mergeItem", "clear", "multiSet", "multiRemove", "multiMerge"
    )

    override fun call(method: String, args: ZynthArgs): JSONObject {
        return when (method) {
            "getItem" -> {
                val key = args.getString("key")
                resultResponse(storage.getItem(key))
            }
            "setItem" -> {
                val key = args.getString("key")
                val value = args.getString("value")
                storage.setItem(key, value)
                resultResponse(null)
            }
            "removeItem" -> {
                val key = args.getString("key")
                storage.removeItem(key)
                resultResponse(null)
            }
            "mergeItem" -> {
                val key = args.getString("key")
                val value = args.getString("value")
                storage.mergeItem(key, value)
                resultResponse(null)
            }
            "clear" -> {
                storage.clear()
                resultResponse(null)
            }
            "getAllKeys" -> resultResponse(storage.getAllKeys())
            "multiGet" -> {
                val keys = args.getList("keys").mapNotNull { it as? String }
                resultResponse(storage.multiGet(keys))
            }
            "multiSet" -> {
                val pairs = parsePairs(args.getList("pairs"))
                storage.multiSet(pairs)
                resultResponse(null)
            }
            "multiRemove" -> {
                val keys = args.getList("keys").mapNotNull { it as? String }
                storage.multiRemove(keys)
                resultResponse(null)
            }
            "multiMerge" -> {
                val pairs = parsePairs(args.getList("pairs"))
                storage.multiMerge(pairs)
                resultResponse(null)
            }
            else -> throw IllegalArgumentException("Unsupported method: $method")
        }
    }

    override fun callSync(method: String, args: ZynthArgs): Any? {
        return when (method) {
            "getItem" -> storage.getItem(args.getString("key"))
            "setItem" -> {
                storage.setItem(args.getString("key"), args.getString("value"))
                null
            }
            "removeItem" -> {
                storage.removeItem(args.getString("key"))
                null
            }
            "mergeItem" -> {
                storage.mergeItem(args.getString("key"), args.getString("value"))
                null
            }
            "clear" -> {
                storage.clear()
                null
            }
            "getAllKeys" -> storage.getAllKeys()
            "multiGet" -> {
                storage.multiGet(args.getList("keys").mapNotNull { it as? String })
            }
            "multiSet" -> {
                storage.multiSet(parsePairs(args.getList("pairs")))
                null
            }
            "multiRemove" -> {
                storage.multiRemove(args.getList("keys").mapNotNull { it as? String })
                null
            }
            "multiMerge" -> {
                storage.multiMerge(parsePairs(args.getList("pairs")))
                null
            }
            else -> throw IllegalArgumentException("Unsupported method: $method")
        }
    }

    private fun parsePairs(value: List<Any?>): List<Pair<String, String>> {
        val result = mutableListOf<Pair<String, String>>()
        for (entry in value) {
            val pair = parsePair(entry)
            if (pair != null) {
                result.add(pair)
            }
        }
        return result
    }

    private fun parsePair(value: Any?): Pair<String, String>? {
        return when (value) {
            is JSONArray -> {
                if (value.length() >= 2) {
                    val key = value.opt(0)
                    val valValue = value.opt(1)
                    if (key is String && valValue is String) {
                        key to valValue
                    } else {
                        null
                    }
                } else {
                    null
                }
            }
            is Array<*> -> {
                if (value.size >= 2) {
                    val key = value[0] as? String
                    val valValue = value[1] as? String
                    if (key != null && valValue != null) {
                        key to valValue
                    } else {
                        null
                    }
                } else {
                    null
                }
            }
            is List<*> -> {
                if (value.size >= 2) {
                    val key = value[0] as? String
                    val valValue = value[1] as? String
                    if (key != null && valValue != null) {
                        key to valValue
                    } else {
                        null
                    }
                } else {
                    null
                }
            }
            is JSONObject -> {
                val key = value.opt("key") as? String
                val valValue = value.opt("value") as? String
                if (key != null && valValue != null) {
                    key to valValue
                } else {
                    null
                }
            }
            is Map<*, *> -> {
                val key = value["key"] as? String
                val valValue = value["value"] as? String
                if (key != null && valValue != null) {
                    key to valValue
                } else {
                    null
                }
            }
            else -> null
        }
    }

    private fun resultResponse(result: Any?): JSONObject {
        return JSONObject().apply {
            put("result", result ?: JSONObject.NULL)
        }
    }
}

private class ZynthAsyncStorageStore(context: Context) {
    private val prefs: SharedPreferences =
        context.getSharedPreferences("zynth_async_storage", Context.MODE_PRIVATE)

    fun getItem(key: String?): String? {
        if (key == null) return null
        return prefs.getString(key, null)
    }

    fun setItem(key: String, value: String) {
        if (!prefs.edit().putString(key, value).commit()) {
            throw java.io.IOException("Failed to persist item to storage")
        }
    }

    fun removeItem(key: String) {
        if (!prefs.edit().remove(key).commit()) {
            throw java.io.IOException("Failed to remove item from storage")
        }
    }

    fun mergeItem(key: String, value: String) {
        val existing = prefs.getString(key, null)
        val merged = mergeJson(existing, value)
        if (!prefs.edit().putString(key, merged).commit()) {
            throw java.io.IOException("Failed to merge item in storage")
        }
    }

    fun clear() {
        if (!prefs.edit().clear().commit()) {
            throw java.io.IOException("Failed to clear storage")
        }
    }

    fun getAllKeys(): JSONArray {
        val keys = prefs.all.keys
        val array = JSONArray()
        for (key in keys) {
            array.put(key)
        }
        return array
    }

    fun multiGet(keys: List<String>): JSONArray {
        val result = JSONArray()
        for (key in keys) {
            val entry = JSONArray()
            entry.put(key)
            entry.put(prefs.getString(key, null))
            result.put(entry)
        }
        return result
    }

    fun multiSet(pairs: List<Pair<String, String>>) {
        val editor = prefs.edit()
        for ((key, value) in pairs) {
            editor.putString(key, value)
        }
        if (!editor.commit()) {
            throw java.io.IOException("Failed to persist multiple items to storage")
        }
    }

    fun multiRemove(keys: List<String>) {
        val editor = prefs.edit()
        for (key in keys) {
            editor.remove(key)
        }
        if (!editor.commit()) {
            throw java.io.IOException("Failed to remove multiple items from storage")
        }
    }

    fun multiMerge(pairs: List<Pair<String, String>>) {
        val editor = prefs.edit()
        for ((key, value) in pairs) {
            val existing = prefs.getString(key, null)
            val merged = mergeJson(existing, value)
            editor.putString(key, merged)
        }
        if (!editor.commit()) {
            throw java.io.IOException("Failed to merge multiple items in storage")
        }
    }

    private fun mergeJson(existing: String?, update: String): String {
        val updateObject = parseJsonObject(update) ?: return update
        val existingObject = parseJsonObject(existing) ?: return update
        val merged = deepMerge(existingObject, updateObject)
        return merged.toString()
    }

    private fun parseJsonObject(value: String?): JSONObject? {
        if (value == null) return null
        return try {
            JSONObject(value)
        } catch (_: Exception) {
            null
        }
    }

    private fun deepMerge(base: JSONObject, update: JSONObject): JSONObject {
        val result = JSONObject(base.toString())
        val keys = update.keys()
        while (keys.hasNext()) {
            val key = keys.next()
            val updateValue = update.opt(key)
            val baseValue = result.opt(key)
            if (updateValue is JSONObject && baseValue is JSONObject) {
                result.put(key, deepMerge(baseValue, updateValue))
            } else {
                result.put(key, updateValue)
            }
        }
        return result
    }
}

File: packages/zynth-async-storage/ios/ZynthAsyncStorageModule.swift
//
//  ZynthAsyncStorageModule.swift
//  ZynthAsyncStorage
//
//  Native AsyncStorage implementation backed by UserDefaults
//

import Foundation
import ZynthKit

@objc(ZynthAsyncStorageModule)
final class ZynthAsyncStorageModule: NSObject, ZynthModule, ZynthSyncModule {
  let name: String = "ZynthAsyncStorage"
  private let storage = ZynthAsyncStorageStore()

  var exportedMethods: [String] {
    return [
      "getItem", "setItem", "removeItem", "mergeItem", "clear",
      "getAllKeys", "multiGet", "multiSet", "multiRemove", "multiMerge"
    ]
  }

  var protectedMethods: [String] {
    return ["setItem", "removeItem", "mergeItem", "clear", "multiSet", "multiRemove", "multiMerge"]
  }

  func call(method: String, args: ZynthArgs) throws -> Any? {
    return try handle(method: method, args: args)
  }

  func callSync(method: String, args: ZynthArgs) throws -> Any? {
    return try handle(method: method, args: args)
  }

  private func handle(method: String, args: ZynthArgs) throws -> Any? {
    switch method {
    case "getItem":
      let key = try args.string("key")
      return ["result": storage.getItem(key) as Any? ?? NSNull()]
    case "setItem":
      let key = try args.string("key")
      let value = try args.string("value")
      storage.setItem(key, value: value)
      return ["result": true]
    case "removeItem":
      let key = try args.string("key")
      storage.removeItem(key)
      return ["result": true]
    case "mergeItem":
      let key = try args.string("key")
      let value = try args.string("value")
      storage.mergeItem(key, value: value)
      return ["result": true]
    case "clear":
      storage.clear()
      return ["result": true]
    case "getAllKeys":
      return ["result": storage.getAllKeys()]
    case "multiGet":
      let keys = try args.array("keys").compactMap { $0 as? String }
      return ["result": storage.multiGet(keys)]
    case "multiSet":
      let pairs = parsePairs(try args.array("pairs"))
      storage.multiSet(pairs)
      return ["result": true]
    case "multiRemove":
      let keys = try args.array("keys").compactMap { $0 as? String }
      storage.multiRemove(keys)
      return ["result": true]
    case "multiMerge":
      let pairs = parsePairs(try args.array("pairs"))
      storage.multiMerge(pairs)
      return ["result": true]
    default:
      throw ZynthModuleError.methodNotExported(module: name, method: method)
    }
  }

  private func parsePairs(_ value: [Any]) -> [(String, String)] {
    var result: [(String, String)] = []
    for entry in value {
      if let pair = parsePair(entry) {
        result.append(pair)
      }
    }
    return result
  }

  private func parsePair(_ value: Any?) -> (String, String)? {
    if let array = value as? [Any], array.count >= 2,
       let key = array[0] as? String,
       let val = array[1] as? String {
      return (key, val)
    }
    if let dict = value as? [String: Any],
       let key = dict["key"] as? String,
       let val = dict["value"] as? String {
      return (key, val)
    }
    return nil
  }
}

final class ZynthAsyncStorageStore {
  private let defaults: UserDefaults
  private let queue = DispatchQueue(label: "dev.zynth.asyncstorage.queue", qos: .utility)
  private let prefix = "zynth.asyncstorage."

  init(defaults: UserDefaults = .standard) {
    self.defaults = defaults
  }

  func getItem(_ key: String) -> String? {
    return queue.sync {
      defaults.string(forKey: prefixed(key))
    }
  }

  func setItem(_ key: String, value: String) {
    queue.sync {
      defaults.set(value, forKey: prefixed(key))
    }
  }

  func removeItem(_ key: String) {
    queue.sync {
      defaults.removeObject(forKey: prefixed(key))
    }
  }

  func mergeItem(_ key: String, value: String) {
    queue.sync {
      let existing = defaults.string(forKey: prefixed(key))
      let merged = mergeJson(existing: existing, update: value)
      defaults.set(merged, forKey: prefixed(key))
    }
  }

  func clear() {
    queue.sync {
      for key in defaults.dictionaryRepresentation().keys where key.hasPrefix(prefix) {
        defaults.removeObject(forKey: key)
      }
    }
  }

  func getAllKeys() -> [String] {
    return queue.sync {
      defaults.dictionaryRepresentation().keys
        .filter { $0.hasPrefix(prefix) }
        .map { String($0.dropFirst(prefix.count)) }
    }
  }

  func multiGet(_ keys: [String]) -> [[Any]] {
    return queue.sync {
      keys.map { key in
        let value = defaults.string(forKey: prefixed(key))
        return [key, value ?? NSNull()]
      }
    }
  }

  func multiSet(_ pairs: [(String, String)]) {
    queue.sync {
      for (key, value) in pairs {
        defaults.set(value, forKey: prefixed(key))
      }
    }
  }

  func multiRemove(_ keys: [String]) {
    queue.sync {
      for key in keys {
        defaults.removeObject(forKey: prefixed(key))
      }
    }
  }

  func multiMerge(_ pairs: [(String, String)]) {
    queue.sync {
      for (key, value) in pairs {
        let existing = defaults.string(forKey: prefixed(key))
        let merged = mergeJson(existing: existing, update: value)
        defaults.set(merged, forKey: prefixed(key))
      }
    }
  }

  private func prefixed(_ key: String) -> String {
    return prefix + key
  }

  private func mergeJson(existing: String?, update: String) -> String {
    guard let updateData = update.data(using: .utf8),
          let updateObject = try? JSONSerialization.jsonObject(with: updateData, options: [])
    else {
      return update
    }

    guard let updateDict = updateObject as? [String: Any] else {
      return update
    }

    guard let existing = existing,
          let existingData = existing.data(using: .utf8),
          let existingObject = try? JSONSerialization.jsonObject(with: existingData, options: []),
          let existingDict = existingObject as? [String: Any]
    else {
      return update
    }

    let merged = deepMerge(base: existingDict, update: updateDict)
    guard let data = try? JSONSerialization.data(withJSONObject: merged, options: [])
    else {
      return update
    }
    return String(data: data, encoding: .utf8) ?? update
  }

  private func deepMerge(base: [String: Any], update: [String: Any]) -> [String: Any] {
    var result = base
    for (key, value) in update {
      if let baseDict = result[key] as? [String: Any],
         let updateDict = value as? [String: Any] {
        result[key] = deepMerge(base: baseDict, update: updateDict)
      } else {
        result[key] = value
      }
    }
    return result
  }
}
